File: scripts/lib/promax_module_split.py
#!/usr/bin/env python3
"""Split mixed Surge modules into ad-only extracts for PROMAX."""

from __future__ import annotations

import os
import re
from typing import Dict, List, Optional, Tuple

from lib.promax_line_classifier import should_keep_promax_line

SECTION_HEADER_RE = re.compile(r"^\[([^\]]+)\]\s*$")


def split_module_sections(
    text: str, *, source_path: Optional[str] = None
) -> Tuple[Dict[str, List[str]], Dict[str, int]]:
    """
    Return (ad_sections, stats).
    stats keys: total_lines, kept_lines, skipped_lines
    """
    ad_sections: Dict[str, List[str]] = {}
    current_section: Optional[str] = None
    total = kept = skipped = 0

    for raw in text.splitlines():
        stripped = raw.strip()
        m = SECTION_HEADER_RE.match(stripped)
        if m and not stripped.startswith("#"):
            current_section = m.group(1)
            continue
        if not stripped or stripped.startswith("#"):
            continue
        if current_section is None:
            continue
        if current_section.lower() in ("general", "ponte"):
            skipped += 1
            total += 1
            continue
        total += 1
        if should_keep_promax_line(stripped, current_section, source_path=source_path):
            ad_sections.setdefault(current_section, []).append(stripped)
            kept += 1
        else:
            skipped += 1

    return ad_sections, {
        "total_lines": total,
        "kept_lines": kept,
        "skipped_lines": skipped,
    }


def format_split_module(
    source_path: str,
    ad_sections: Dict[str, List[str]],
    *,
    note: str,
) -> str:
    """Minimal sgmodule fragment for review under module/build/promax_splits/."""
    base = os.path.basename(source_path)
    lines = [
        f"#!name=PROMAX split · {base}",
        f"#!desc={note}",
        "#!author=ScriptHub-Automated",
        f"#!split-source={source_path}",
    ]
    order = (
        "Rule",
        "URL Rewrite",
        "Map Local",
        "Script",
        "Body Rewrite",
        "Header Rewrite",
        "MITM",
    )
    seen = set(ad_sections)
    for sec in order:
        if sec not in ad_sections:
            continue
        body = ad_sections[sec]
        if not body:
            continue
        lines.append("")
        lines.append(f"[{sec}]")
        lines.extend(body)
    for sec in sorted(seen):
        if sec in order:
            continue
        lines.append("")
        lines.append(f"[{sec}]")
        lines.extend(ad_sections[sec])
    lines.append("")
    return "\n".join(lines)

File: go_scripts/pkg/hub/promax_module_split.go
package hub

import (
	"fmt"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var sectionHeaderRe = regexp.MustCompile(`^\[([^\]]+)\]\s*$`)

type SplitStats struct {
	TotalLines   int
	KeptLines    int
	SkippedLines int
}

func SplitModuleSections(text string, sourcePath string) (map[string][]string, SplitStats) {
	adSections := make(map[string][]string)
	currentSection := ""
	total := 0
	kept := 0
	skipped := 0

	lines := strings.Split(text, "\n")
	for _, raw := range lines {
		stripped := strings.TrimSpace(raw)
		m := sectionHeaderRe.FindStringSubmatch(stripped)
		if m != nil && !strings.HasPrefix(stripped, "#") {
			currentSection = m[1]
			continue
		}
		if stripped == "" || strings.HasPrefix(stripped, "#") {
			continue
		}
		if currentSection == "" {
			continue
		}
		secLower := strings.ToLower(currentSection)
		if secLower == "general" || secLower == "ponte" {
			skipped++
			total++
			continue
		}
		total++
		if ShouldKeepPromaxLine(stripped, currentSection, sourcePath) {
			adSections[currentSection] = append(adSections[currentSection], stripped)
			kept++
		} else {
			skipped++
		}
	}

	return adSections, SplitStats{
		TotalLines:   total,
		KeptLines:    kept,
		SkippedLines: skipped,
	}
}

func FormatSplitModule(sourcePath string, adSections map[string][]string, note string) string {
	base := filepath.Base(sourcePath)
	lines := []string{
		fmt.Sprintf("#!name=PROMAX split · %s", base),
		fmt.Sprintf("#!desc=%s", note),
		"#!author=ScriptHub-Automated",
		fmt.Sprintf("#!split-source=%s", sourcePath),
	}

	order := []string{
		"Rule",
		"URL Rewrite",
		"Map Local",
		"Script",
		"Body Rewrite",
		"Header Rewrite",
		"MITM",
	}

	seen := make(map[string]bool)
	for k := range adSections {
		seen[k] = true
	}

	for _, sec := range order {
		if !seen[sec] {
			continue
		}
		body := adSections[sec]
		if len(body) == 0 {
			continue
		}
		lines = append(lines, "")
		lines = append(lines, fmt.Sprintf("[%s]", sec))
		lines = append(lines, body...)
	}

	var remaining []string
	for sec := range seen {
		inOrder := false
		for _, o := range order {
			if o == sec {
				inOrder = true
				break
			}
		}
		if !inOrder {
			remaining = append(remaining, sec)
		}
	}
	sort.Strings(remaining)

	for _, sec := range remaining {
		lines = append(lines, "")
		lines = append(lines, fmt.Sprintf("[%s]", sec))
		lines = append(lines, adSections[sec]...)
	}
	lines = append(lines, "")

	return strings.Join(lines, "\n")
}
